examples/Summarizing_with_controllable_detail.ipynb

```diff
@@ -0,0 +1,657 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import openai\n",
+    "import tiktoken\n",
+    "from typing import List, Tuple, Optional\n",
+    "from tqdm import tqdm"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# open dataset containing part of the text of the Wikipedia page for the United States\n",
+    "with open(\"data/united_states_wikipedia.txt\", \"r\") as file:\n",
+    "    united_states_wikipedia_text = file.read()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": "15781"
+     },
+     "execution_count": 3,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "# load encoding and check the length of dataset\n",
+    "encoding = tiktoken.encoding_for_model('gpt-3.5-turbo')\n",
+    "len(encoding.encode(united_states_wikipedia_text))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We'll define a simple utility to wrap calls to the OpenAI API."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def get_chat_completion(messages, model='gpt-3.5-turbo'):\n",
+    "    response = openai.ChatCompletion.create(\n",
+    "        model=model,\n",
+    "        messages=messages,\n",
+    "        temperature=0,\n",
+    "    )\n",
+    "    return response.choices[0].message['content']"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Next we'll define some utilities to chunk a large document into smaller pieces."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def tokenize(text: str) -> List[str]:\n",
+    "    encoding = tiktoken.encoding_for_model('gpt-3.5-turbo')\n",
+    "    return encoding.encode(text)\n",
+    "\n",
+    "\n",
+    "def chunk_on_delimiter(input_string, max_tokens, delimiter):\n",
+    "    chunks = input_string.split(delimiter)\n",
+    "    combined_chunks, _, dropped_chunk_count = combine_chunks_with_no_minimum(\n",
+    "        chunks, max_tokens, chunk_delimiter=delimiter, add_ellipsis_for_overflow=True\n",
+    "    )\n",
+    "    if dropped_chunk_count > 0:\n",
+    "        print(f\"warning: {dropped_chunk_count} chunks were dropped due to overflow\")\n",
+    "    combined_chunks = [f\"{chunk}{delimiter}\" for chunk in combined_chunks]\n",
+    "    return combined_chunks\n",
+    "\n",
+    "\n",
+    "def combine_chunks_with_no_minimum(\n",
```

Comment:
Might be useful to introduce these functions at a high level, and maybe also how they fit together. (So a skimming reader can still follow along).